components/ws-manager/pkg/manager/monitor.go

```diff
@@ -1093,10 +1093,22 @@ func (m *Monitor) finalizeWorkspaceContent(ctx context.Context, wso *workspaceOb
 		// only build prebuild snapshots of initialized/ready workspaces.
 		if tpe == api.WorkspaceType_PREBUILD {
 			_, err = snc.WaitForInit(ctx, &wsdaemon.WaitForInitRequest{Id: workspaceID})
-			if st, ok := grpc_status.FromError(err); ok && st.Code() == codes.FailedPrecondition &&
-				(st.Message() == "workspace is not initializing or ready" || st.Message() == "workspace is not ready") {
-				log.Warn("skipping snapshot creation because content-initializer never finished or the workspace reached a ready state")
-				doSnapshot = false
+			if st, ok := grpc_status.FromError(err); ok {
+				if st.Code() == codes.FailedPrecondition &&
+					(st.Message() == "workspace is not initializing or ready" || st.Message() == "workspace is not ready") {
+					log.Warn("skipping snapshot creation because content-initializer never finished or the workspace reached a ready state")
+					doSnapshot = false
+				} else if st.Code() == codes.NotFound {
+					// the workspace has gone some reason
+					// e.g. since it was a retry, it already succeeded the first time.
+					log.WithError(err).Warnf("skipping snapshot and disposing because the workspace has already gone")
+					val, ok := m.finalizerMap.LoadAndDelete(workspaceID)
+					if ok {
+						cancelReq := val.(context.CancelFunc)
+						cancelReq()
+					}
+					return false, &csapi.GitStatus{}, nil
+				}
```

Comment:
Can we move these lines https://github.com/gitpod-io/gitpod/blob/b8669f351eb66bac37b4a06a6753b4390232cba6/components/ws-manager/pkg/manager/monitor.go#L1115-L1126 upper?
Therefore, we can reuse the defer func.

Reply:
@jenting Thanks for your feedback. fixed